presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -125,57 +127,113 @@ private static Expression toPredicate(Domain domain, QualifiedNameReference refe
+        checkState(!rangeConjuncts.isEmpty());
+        return combineConjuncts(rangeConjuncts);
+    }
+
+    private static Expression combineRangeWithExcludedPoints(Type type, QualifiedNameReference reference, Range range, List<Expression> excludedPoints)
+    {
+        checkState(!(excludedPoints.isEmpty() && range.isAll()), "cannot combine empty excluded points list with an ALL range");
+
+        if (!range.isAll() && excludedPoints.isEmpty()) {
+            return processRange(type, range, reference);
+        }
+
+        if (!excludedPoints.isEmpty()) {
+            Expression excludedPointsExpression = new InPredicate(reference, new InListExpression(excludedPoints));
+            if (excludedPoints.size() == 1) {
+                excludedPointsExpression = new ComparisonExpression(EQUAL, reference, getOnlyElement(excludedPoints));
+            }
+
+            if (range.isAll()) {
+                return new NotExpression(excludedPointsExpression);
+            }
+            else {
+                return and(processRange(type, range, reference), new NotExpression(excludedPointsExpression));
+            }
+        }
+
+        return processRange(type, range, reference);
+    }
+
     private static List<Expression> extractDisjuncts(Type type, Ranges ranges, QualifiedNameReference reference)
     {
         List<Expression> disjuncts = new ArrayList<>();
         List<Expression> singleValues = new ArrayList<>();
-        for (Range range : ranges.getOrderedRanges()) {
+        List<Range> orderedRanges = ranges.getOrderedRanges();
+
+        // attempt to optimize ranges that can be coalesced as long as certain points are excluded
+        SortedRangeSet sortedRangeSet = SortedRangeSet.copyOf(type, orderedRanges);
+        SortedRangeSet complement = sortedRangeSet.complement();
+        List<Range> singleValuesInComplement = complement.getOrderedRanges().stream().filter(Range::isSingleValue).collect(toList());
+        List<Range> originalUnionSingleValues = SortedRangeSet.copyOf(type, singleValuesInComplement).union(sortedRangeSet).getOrderedRanges();
+
+        int i = 0;
+        for (Range range : originalUnionSingleValues) {
+            List<Expression> singleValuesInRange = new ArrayList<>();
+
+            while (i < singleValuesInComplement.size() && range.contains(singleValuesInComplement.get(i))) {
+                singleValuesInRange.add(toExpression(singleValuesInComplement.get(i).getSingleValue(), type));
+                i++;
+            }
+
+            // if only a single point is in the exclusions list, this optimization is not worth it
+            if (singleValuesInRange.size() > 1) {
+                disjuncts.add(combineRangeWithExcludedPoints(type, reference, range, singleValuesInRange));
+                orderedRanges = orderedRanges.stream().filter(r -> !range.overlaps(r)).collect(Collectors.toList());
```

Comment:
merge this into the lop below. it should be simpler to not handle this case specially.